bpf-gpl/xdp.c

```diff
@@ -84,10 +84,18 @@ static CALI_BPF_INLINE int calico_xdp(struct xdp_md *xdp)
 		goto allow;
 	}
 
-	// Allow a packet if it hits an entry in the failsafe map
+	// Allow a packet if it hits an entry in the inbound ports failsafe map
 	if (is_failsafe_in(ctx.state->ip_proto, ctx.state->dport, ctx.state->ip_src)) {
 		CALI_DEBUG("Inbound failsafe port: %d. Skip policy\n", ctx.state->dport);
-		goto allow_with_metadata;
+		ctx.state->pol_rc = CALI_POL_ALLOW;
+		goto allow;
+	}
+
+	// Allow a packet if it hits an entry in the outbound ports failsafe map
+	if (is_failsafe_out(ctx.state->ip_proto, ctx.state->sport, ctx.state->ip_src)) {
+		CALI_DEBUG("Outbound failsafe port: %d. Skip policy\n", ctx.state->sport);
+		ctx.state->pol_rc = CALI_POL_ALLOW;
+		goto allow;
```

Comment:
the XDP program runs only in RX path. I am not totally sure about how the failsafe works, but is checking "is_failsafe_out" valid for RX path?

Reply:
one more thing, since after matching against failsafe map we return XDP_PASS and leave XDP execution, do we need to set "state.pol_rc" field?

Reply:
Thanks @mazdakn , it definitely makes sense to add a comment about why this code is here, so I will do that.

Reply:
@fasaxc The "goto allow" here is in fact just a fall through to TC processing.  It does not do the marking that would cause the TC program to exit early.

Reply:
Ah, got it. This is a tricky one!